AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/HierarchicalVersionUtils.dfy

```diff
@@ -0,0 +1,128 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+include "../Model/AwsCryptographyKeyStoreTypes.dfy"
+include "Structure.dfy"
+include "KMSKeystoreOperations.dfy"
+include "KmsArn.dfy"
+
+module HierarchicalVersionUtils {
+
+  import opened StandardLibrary
+  import opened Wrappers
+  import opened Seq
+
+  import ErrorMessages = KeyStoreErrorMessages
+  import Types = AwsCryptographyKeyStoreTypes
+  import KMS = ComAmazonawsKmsTypes
+  
+  import AtomicPrimitives
+  import UTF8
+  import KMSKeystoreOperations
+  import KmsArn
+  import Structure
+  
+  function method GetMdDigestFromEC(
+    item: Types.EncryptionContextString
+  ) : (output: Types.EncryptionContextString)
+  ensures output.Keys == item.Keys - {Structure.TABLE_FIELD}
+  ensures forall k :: k in output ==> output[k] == item[k]
+  ensures forall k :: k in output ==> k !in {Structure.TABLE_FIELD}
+  {
+    map k <- item.Keys - {Structure.TABLE_FIELD}
+      :: k := item[k]
+  }
+  
+  method GetHV2EC(
+    ecStringMap: Types.EncryptionContextString
+  ) returns (output: Types.EncryptionContextString)
+    ensures Structure.Hv2EncryptionContext?(output)
+  {
+    var withoutRestrictedField := RemoveRestrictedFields(ecStringMap);
+    var items := withoutRestrictedField.Items;
+    var newMap: map<string, string> := map[];
+
+    while items != {}
+        decreases |items|
+    {
+        var item :| item in items;
+        items := items - { item };
+        if (|item.0| >= |Structure.ENCRYPTION_CONTEXT_PREFIX| && item.0[..|Structure.ENCRYPTION_CONTEXT_PREFIX|] == Structure.ENCRYPTION_CONTEXT_PREFIX) {
+          var newKey := item.0[|Structure.ENCRYPTION_CONTEXT_PREFIX|..];
+          newMap := newMap[newKey := item.1];
+        } else {
+          newMap := newMap[item.0 := item.1];
+        }
+    }
+    return withoutRestrictedField;
+  }
+
+  function method RemoveRestrictedFields(a:map<string, string>) : (output:map<string, string>)
+    ensures Structure.Hv2EncryptionContext?(output)
+  {
+    a - Structure.BRANCH_KEY_RESTRICTED_FIELD_NAMES
+  }
```

Comment:
Nit: I hate bad variable names. I get it, `input` vs `a` really does not matter. 
But some part of me just does not have the flexibility to accept this.
```suggestion
  function method RemoveRestrictedFields(input:map<string, string>) : (output:map<string, string>)
    ensures Structure.Hv2EncryptionContext?(output)
  {
    input - Structure.BRANCH_KEY_RESTRICTED_FIELD_NAMES
  }
```